### PR TITLE
[DiscordCoreAPI] Update to version 2022-06-18.

### PR DESCRIPTION
**Pull request by RealTimeChris**
**Describe the pull request**

- #### What does your PR fix?
It adds support for Discord's latest API endpoints. I also refactored some of the core architecture - instead of using 1 thread per connection in the instance of sharding, it utilizes a threadpool now. I've also: reworked some of the exception handling, improved the cmake build system to handle itself more effectively, and removed the need for 2 of the dependencies.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  <Linux, Windows>, Yes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes.

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
 Yes I have.

### CODE REVIEW

File: ports/discordcoreapi/portfile.cmake [hunk 1]
```diff
@@ -16,16 +16,8 @@ vcpkg_cmake_configure(
 
 vcpkg_cmake_install()
 
-vcpkg_cmake_config_fixup(NO_PREFIX_CORRECTION)
```
Comment (JackBoosY) [inline ports/discordcoreapi/portfile.cmake:19]: Why remove this? Does the upstream no longer provide to generate the cmake configure files?

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/discordcoreapi/portfile.cmake (hunk 1 above)
Comment (JackBoosY) [reply ports/discordcoreapi/portfile.cmake:19]: I don't get it. Even if the *config.cmake is generated correctly, we also need to use `vcpkg_cmake_config_fixup()` with no params here.

Comment (RealTimeChris) [reply ports/discordcoreapi/portfile.cmake:19]: It seems to build fine on the Windows and Linux configurations I've tested it on without it - but if it is the case that it is required then I have no problems putting it back.

Comment (JackBoosY) [reply ports/discordcoreapi/portfile.cmake:19]: We need to ensure that the absolute path is changed to relative path to support feature export and binarycaching, and the debug configuration should exist in the same directory as the release configuration.

Comment (RealTimeChris) [reply ports/discordcoreapi/portfile.cmake:19]: I'm fairly certain that this is the case, the way I have the cmake script set up.

**Review by JackBoosY — APPROVED** [review]

File: versions/d-/discordcoreapi.json [hunk 2]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "ad04fb27ce70ec41f7952a752f74831e20c49bd4",
```
Comment (JackBoosY) [inline versions/d-/discordcoreapi.json:4]:
```suggestion
      "git-tree": "c452db017823a155eec5eba2a0dca0d21f12d541",
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
The upstream cmake seems has some bugs:
In debug:
```
-- Installing: D:/packages/discordcoreapi_x64-uwp/share/discordcoreapi/DiscordCoreAPIConfig.cmake
-- Installing: D:/packages/discordcoreapi_x64-uwp/share/discordcoreapi/DiscordCoreAPIConfigVersion.cmake
...
-- Installing: D:/packages/discordcoreapi_x64-uwp/share/discordcoreapi/DiscordCoreAPITargets.cmake
-- Installing: D:/packages/discordcoreapi_x64-uwp/share/discordcoreapi/DiscordCoreAPITargets-debug.cmake
```
Theses files should be installed into _D:/packages/discordcoreapi_x64-uwp/debug/share/discordcoreapi_.
The installation code should be:
```cmake
install(EXPORT DiscordCoreAPITargets ... DESTINATION share/discordcoreapi)
install(FILES DiscordCoreAPIConfig.cmake DESINTATION share/discordcoreapi)
install(FILES DiscordCoreAPIConfigVersion.cmake DESINTATION share/discordcoreapi)
```

File: ports/discordcoreapi/portfile.cmake [hunk 3]
```diff
@@ -1,31 +1,31 @@
...
-if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
-    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/bin" "${CURRENT_PACKAGES_DIR}/debug/bin")
-endif()
-
-file(
-	INSTALL "${SOURCE_PATH}/License.md"
-	DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}"
-	RENAME copyright
-)
+if(VCPKG_TARGET_IS_LINUX)
```
Comment (JackBoosY) [inline ports/discordcoreapi/portfile.cmake:1]: Did you convert the file's format?

**Review by JackBoosY — COMMENTED** [review]
Did you test cmake usage?

File: ports/discordcoreapi/portfile.cmake (hunk 3 above)
Comment (RealTimeChris) [reply ports/discordcoreapi/portfile.cmake:1]: What do you mean?

**Comment by RealTimeChris** [comment]
> Did you test cmake usage?

Yes I did - it works as intended.

**Review by JackBoosY — APPROVED** [review]

**Pull request merged**